rosbag2_storage/src/rosbag2_storage/impl/storage_factory_impl.hpp

```diff
@@ -43,6 +43,59 @@ get_class_loader()
   return std::make_shared<pluginlib::ClassLoader<InterfaceT>>("rosbag2_storage", lookup_name);
 }
 
+template<typename InterfaceT>
+std::shared_ptr<InterfaceT>
+try_load_plugin(
+  std::shared_ptr<pluginlib::ClassLoader<InterfaceT>> class_loader,
+  const std::string & plugin_name)
+{
+  std::shared_ptr<InterfaceT> instance;
+  try {
+    auto unmanaged_instance = class_loader->createUnmanagedInstance(plugin_name);
+    instance = std::shared_ptr<InterfaceT>(unmanaged_instance);
+  } catch (const std::runtime_error & ex) {
+    ROSBAG2_STORAGE_LOG_ERROR_STREAM(
+      "Unable to load plugin: " << ex.what());
+  }
+  return instance;
+}
+
+
+template<
+  typename InterfaceT,
+  storage_interfaces::IOFlag flag = StorageTraits<InterfaceT>::io_flag
+>
+std::shared_ptr<InterfaceT>
+try_detect_and_open_storage(
+  std::shared_ptr<pluginlib::ClassLoader<InterfaceT>> class_loader,
+  const StorageOptions & storage_options)
+{
+  bool creating_file = flag != storage_interfaces::IOFlag::READ_ONLY;
+  if (creating_file) {
+    ROSBAG2_STORAGE_LOG_ERROR("Can not auto-choose storage for writing.");
+    return nullptr;
+  }
+
+  const auto & registered_classes = class_loader->getDeclaredClasses();
+  for (const auto & registered_class : registered_classes) {
+    std::shared_ptr<InterfaceT> instance = try_load_plugin(class_loader, registered_class);
+    if (instance == nullptr) {
+      continue;
+    }
+    ROSBAG2_STORAGE_LOG_DEBUG_STREAM(
+      "Trying storage implementation '" << registered_class << "'.");
+    try {
+      instance->open(storage_options, flag);
+      ROSBAG2_STORAGE_LOG_DEBUG_STREAM(
+        "Success, using implementation '" << registered_class << "'.");
+      return instance;
+    } catch (const std::runtime_error & /* ex */) {
+      continue;
```

Comment:
I would suggest to add logging in debug stream about failure on opening `registered_class` i.e. plugin name

Reply:
This line is "if we succeeded to open a plugin, but it failed to open the file" - this is expected to happen every time. E.g. "try sqlite - open? fail. continue. try mcap - open? success. return"

Reply:
Yahh, every time we trying to load and open with fail when storage ID is not specified explicitly. But at least we will see in debug stream which plugins we have tried and on what stage we failed.